Support LXA TAC Gen 2 devices (BSP)

hnez opened the pull request:
Hardware generations
-----------------------------

There is a new batch of LXA TAC devices with a slightly different set of hardware features (**Gen 2**).
Here is an overview of the current hardware generations:

- **Gen 1 Stock**
  - 128px x 64px OLED screen
  - No LED indicators on the top side of the device.
- **Gen 1 LCD**
  - Like **Gen 1 Stock** but retrofitted with a 240px x 240px LCD screen
  - All of the **Gen 1 Stock** devices should be modified to this state
- **Gen 2**
  - 240px x 240px LCD screen
  - LED indicators on the top side of the device

With this PR applied only hardware generations **Gen 1 LCD** and **Gen 2** are supported. Support for **Gen 1 Stock** is dropped.

Related pull requests
----------------------------

- `meta-lxatac`: This PR continues the work started in #10 to support **Gen 1 LCD** and it will thus be closed. There is however an open tasks there that have not yet been addressed:
  - [x] The panel-mipi-dbi recipe should be split into two recpies, one for the tool and one that uses the tool to build the firmware.
- `tacd`: `tacd`s with the PR linux-automation/tacd#13 applied depend on the devicetree changes made in this PR to run. Merging of the PRs should be done in tandem. In addition to that:
  - [x] Update the `tacd` recipe to use the upstream repo and `main` branch with the PR applied.
  - [x] Update the `tacd-webinterface` recipe to use the upstream repo and `main` branch with the PR applied.

Other open tasks
-----------------------

- [x] Decide how the RGB status LED on Gen 2 devices should behave during boot. Do we e.g. want:
  1) Set a "in progress" color in barebox and change to an "error" color if something fails.
  2) Set another "in progress" color in linux or start pulsing in the previous color to indicate the bootloader -> linux change.
  3) Set a "sucess" color in linux when e.g. `boot-complete.target` is reached. Or should this be done in the `tacd`?
- [x] Clean up …
- [ ] … and review the barebox device detection and devicetree selection
- [x] Clean up …
- [x] … and review the linux devicetree / defconfig changes.
- [x] This PR also upgrades the linux kernel to `6.3` make sure that is released and the recipe is updated to the released version. Or backport the changes to `6.2`.
- [x] The `tacd.service` currently uses fixed `gpioset` commands to turn off the `DUT` on failure. This does not work with the Gen 2 powerboard which does not have a GPIO expander. Find a different solution.
- [x] When interacting with a Gen 1 device over the debug port it seemed that barebox as become a bit hard of hearing, e.g. missed a lot of characters. Investigate/fix that.
  20230418 lgo: @SmithChart has found out that this is caused by a floating GPIO line for the input button on the powerboard if no power board is connected. The solution is to enable internal pullups for the buttons.

hnez committed:
meta-lxatac-software: tacd{,-webinterface}: Update to recent main

This involves:

- Support for the larger LCD
- Switch IOBus and UART power via regulators instead of GPIOs
- Change GPIO line names for DUT PWR
- Support for the LEDs on Gen 2 boards
- No longer fail when NetworkManager is not available on startup
- Add a setup mode wizard brings an LXA TAC from "fresh out of the box"
  to "mostly usable" in as few steps as possible.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez committed:
meta-lxatac-software: panel-mipi-dbi: add recipe

This python script compiles a list of human readable commands that
set up a MIPI DBI Display down to a firmware blob to be loaded by the
panel-mipi-dbi Linux driver.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez committed:
meta-lxatac-software: panel-mipi-dbi: add setup script for shineworld,lh133k

The config file sets up the display with a 180° rotation, matching the
orientation in the LXA TAC.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez committed:
meta-lxatac-bsp: linux-lxatac: upgrade to 6.3 with Gen 2 support

umpf tag is now: 6.3/customers/lxa/lxatac/20230509-1
The tag includes the following topic branches:

  - v6.3/customers/lxa/lxatac
  - v6.3/topic/stm32mp1-spi
  - v6.3/topic/pwm-stm32
  - v6.3/topic/pwm-backlight

The kernel config supports Gen 1 TACs that have been retrofitted with an
LCD screen and Gen 2 TACs.

This commit adds patches to achieve flicker free splash screen handover
from bootloader to userspace by keeping PWMs and backlight running
and a fix for the DUT UART not blinking.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>
Signed-off-by: Philipp Zabel <p.zabel@pengutronix.de>

hnez committed:
meta-lxatac-bsp: lxatac-led-setup: add script to set up LED triggers

The Gen 2 LXA TAC has a lot of LEDs that were not present on Gen 1 TACs.
Some of these should use triggers that can only be set up via sysfs.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez committed:
meta-lxatac-bsp: barebox: disable OLED driver

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez committed:
meta-lxatac-bsp: barebox: setup RGB status LED

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez committed:
meta-lxatac-software: tacd: delegate DUT power off on failure to a script

The previous ExecStopPost entry stopped working with the new powerboards
without GPIO expander.
Luckily there is no gpiochip10 now, otherwise the command would have
switched around some random GPIOs.
Delegate the failsafe handling to a script that uses gpiofind to find
the correct GPIOs to turn DUT power off.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez committed:
meta-lxatac-bsp: barebox: update to 2023.03.0

This is done in preparation for the upcoming Gen 1 / Gen 2 changes.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez committed:
meta-lxatac-bsp: barebox: upgrade to version with Gen 1 / Gen 2 LXA TAC hardware support

The detection of the LXA TAC Generation happens in the Barebox Pre Bootloader
based on the traces connected to the GPIOZ bank.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez committed:
meta-lxatac-{bsp,software}: rauc: Set a more fitting compatible

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

pH5 committed:
meta-lxatac-bsp: barebox: add LCD panel support

Signed-off-by: Philipp Zabel <p.zabel@pengutronix.de>

pH5 committed:
meta-lxatac-bsp: barebox: add splash screen

Show the splash screen with a red background on boot failure.

Signed-off-by: Philipp Zabel <p.zabel@pengutronix.de>

pH5 committed:
meta-lxatac-bsp: barebox: LCD partial updates to speed up fbconsole

A full screen update takes about 30 ms to transfer via SPI. The
fbconsole flushes the framebuffer with every character. Add simple
damage tracking and partial display updates to speed up fbconsole.

Signed-off-by: Philipp Zabel <p.zabel@pengutronix.de>

pH5 committed:
meta-lxatac-bsp: barebox: add LCD boot and error messages

Now that fbconsole has passable speed, print which root partition
is about to (or has just failed to) be booted on the LCD.

Signed-off-by: Philipp Zabel <p.zabel@pengutronix.de>

SmithChart committed:
README: Update opening procedure

With the Gen 2 devices the easiest way to access the bootmode pins is by
removing the lower shell part containing the power board.
So no need to disconnect the display anymore.

Signed-off-by: Chris Fiege <cfi@pengutronix.de>

hnez committed:
meta-lxatac-bsp: barebox: improved damage tracking and button pullups

Without the internal pullups enabled for the input buttons the line to the
button on the power board is floating when not power board is connected.
This results in barebox handling lots of button events, making it
mostly unresponsive to e.g. serial line input.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

olerem commented on meta-lxatac-bsp/recipes-kernel/linux/files/defconfig:
```diff
@@ -1704,6 +1702,7 @@ CONFIG_NET_VRF=y
 CONFIG_NET_DSA_MICROCHIP_KSZ_COMMON=y
 # CONFIG_NET_DSA_MICROCHIP_KSZ9477_I2C is not set
 CONFIG_NET_DSA_MICROCHIP_KSZ_SPI=y
+CONFIG_NET_DSA_MICROCHIP_KSZ_PTP=y
```
I would not recommend to enable KSZ_PTP. This change will cost about 50Mbits/s
Here are iperf3 results with KSZ_PTP enabled:
[ ID] Interval           Transfer     Bitrate         Retr                                                                                               
[  5]   0.00-100.00 sec  7.09 GBytes   609 Mbits/sec  140             sender   
[  5]   0.00-100.04 sec  7.09 GBytes   609 Mbits/sec                  receiver
Here are iperf3 results with KSZ_PTP disabled:
[ ID] Interval           Transfer     Bitrate         Retr                                                                                               
[  5]   0.00-100.00 sec  7.68 GBytes   660 Mbits/sec  186             sender     
[  5]   0.00-100.04 sec  7.68 GBytes   659 Mbits/sec                  receiver

hnez replied:
Thanks for the tip and measurements. I've disabled the feature in the newest version of this PR.

hnez committed:
meta-lxatac-bsp: genimage/repart: use 2G root partition instead of 1G

We are close to filling our 1Gb root partitions.
Changing the partition layout with more devices in the field will get
harder and harder.
Increase the partition size now to be safe.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez commented on meta-lxatac-bsp/recipes-kernel/linux/files/defconfig:
```diff
@@ -1704,6 +1701,7 @@ CONFIG_NET_VRF=y
 CONFIG_NET_DSA_MICROCHIP_KSZ_COMMON=y
 # CONFIG_NET_DSA_MICROCHIP_KSZ9477_I2C is not set
 CONFIG_NET_DSA_MICROCHIP_KSZ_SPI=y
+# CONFIG_NET_DSA_MICROCHIP_KSZ_PTP is not set
```
Thanks to the tip from @olerem I've now disabled this feature.

jluebbe commented:
I've looked at the Linux defconfig changes and found no issues.
